File: matlab/rifta/height_2_slopes.m
function [Zx, Zy] = height_2_slopes(...
    X,... X coordinates
    Y,... Y coordinates [m]
    Z,... height map [m]
    mpp,... [m/pixel]
    spot_sz... spot size that used to calculate the window size [m]
    )
% Generate the slope maps from height map, mimicking the NSP measurement
% results

%% 0. Get the required sizes
[m, n] = size(Z);   % obtaint he rows and cols of the height map Z
hf_win_sz = round(0.5*spot_sz / mpp);   % half windows size
Zx = zeros(m, n);
Zy = zeros(m, n);

Z = remove_surface(X,Y,Z);

%% 1. Calculate the slope maps by sliding the window
for i = 1 : m
    for j = 1 : n
        rows = i - hf_win_sz : i + hf_win_sz;   % construct rows of the win
        cols = j - hf_win_sz : j + hf_win_sz;   % construct cols of the win
        
        rows = rows(rows>=1 & rows<=m); % ensure rows within limit
        cols = cols(cols>=1 & cols<=n); % ensure cols within limit
        
        Xwin = X(rows, cols);
        Ywin = Y(rows, cols);
        Zwin = Z(rows, cols);   % get the window's heights
        
        [~,~,f] = remove_surface(Xwin, Ywin, Zwin); % calculate the slopes
        Zx(i,j) = f(2); % obtain the slope x
        Zy(i,j) = f(3); % obtain the slope y
    end
end

end

File: python/rifta/height_2_slopes.py
import numpy as np

def height_2_slopes(X,Y,Z, mpp,spot_sz):
    "0.获取所需的尺寸"
    m, n = Z.shape
    hf_win_sz = round(0.5 * spot_sz / mpp)                                #半个窗口大小（round小数四舍五入为整数）
    Zx = np.zeros((m, n))                                                 #生成0矩阵
    Zy = np.zeros((m, n))
    Z,_,_ = remove_surface(X, Y, Z)
    "1.通过滑动窗口计算坡度图"
    for i in np.arange(1, m + 1):  # (arange循环内的代码)
        for j in np.arange(1, n + 1):
            rows = np.arange(i - hf_win_sz, i + hf_win_sz + 1)  # 构建win的行
            cols = np.arange(j - hf_win_sz, j + hf_win_sz + 1)  # 构建win的列
            rows = rows[(rows >= 1) & (rows <= m)]  # 保证rows的范围
            cols = cols[(cols >= 1) & (cols <= n)]  # 保证cols的范围
            Xwin = X[rows[:, np.newaxis] - 1, cols - 1]  # get the window's heights
            Ywin = Y[rows[:, np.newaxis] - 1, cols - 1]  # np.newaxis 是一个用于增加数组维度的索引器。
            Zwin = Z[rows[:, np.newaxis] - 1, cols - 1]
            _, _, f = remove_surface(Xwin, Ywin, Zwin)  # 忽略前两个数值只接受第三个数值
            Zx[i - 1, j - 1] = f[1][0] # 求斜率x
            Zy[i - 1, j - 1] = f[2][0]
    return Zx, Zy